src/sagas/dependency.saga.js

```diff
@@ -0,0 +1,325 @@
+// @flow
+
+import { select, call, put, cancel, takeEvery } from 'redux-saga/effects';
+import { getPathForProjectId } from '../reducers/paths.reducer';
+import {
+  loadProjectDependency,
+  loadProjectDependencies,
+} from '../services/read-from-disk.service';
+import {
+  installDependency,
+  uninstallDependency,
+  installDependencies,
+  uninstallDependencies,
+} from '../services/dependencies.service';
+import {
+  ADD_DEPENDENCY_START,
+  ADD_DEPENDENCY_ERROR,
+  ADD_DEPENDENCY_FINISH,
+  UPDATE_DEPENDENCY_START,
+  UPDATE_DEPENDENCY_ERROR,
+  UPDATE_DEPENDENCY_FINISH,
+  DELETE_DEPENDENCY_START,
+  DELETE_DEPENDENCY_ERROR,
+  DELETE_DEPENDENCY_FINISH,
+  ADD_DEPENDENCIES_START,
+  ADD_DEPENDENCIES_ERROR,
+  ADD_DEPENDENCIES_FINISH,
+  DELETE_DEPENDENCIES_START,
+  DELETE_DEPENDENCIES_ERROR,
+  DELETE_DEPENDENCIES_FINISH,
+  addDependencyFinish,
+  addDependencyError,
+  updateDependencyFinish,
+  updateDependencyError,
+  deleteDependencyFinish,
+  deleteDependencyError,
+  addDependenciesStart,
+  addDependenciesError,
+  addDependenciesFinish,
+  deleteDependenciesStart,
+  deleteDependenciesError,
+  deleteDependenciesFinish,
+  showNotification,
+  updateNotification,
+  completeNotification,
+  failNotification,
+} from '../actions';
+
+// Since redux-saga is intended to handle side-effects of actions,
+// all sagas are run *after* all reducers, even though they are
+// incorporated into redux as a middleware. As such, we cannot rely
+// on a store-based check for whether or not a package.json is locked,
+// as any locking actions will lock it in the store before our saga
+// has a chance to check if it is locked or not. As such, we must handle
+// monitoring the lock status within the saga, so we can control the
+// order of checking.
+//
+// NOTE: the previous reducer added false entries for new projects on
+// addition and for all projects after a refresh, but undefined is
+// falsey and has the same effect so I elected not to add an extra
+// saga for these actions.
+const lock: {
+  [projectId: string]: boolean,
+} = {};
+
+// The queue is split out into subqueues per project. Each subqueue item
+// denotes its action (add/delete), the persistent notification's ID
+// for updating its progress in the UI, and a list of dependencies to
+// be installed/removed by that action.
+const queue: {
+  [projectId: string]: Array<{
+    action: string,
+    notificationId: string,
+    dependencies: Array<{
+      dependencyName: string,
+      version?: string,
+      updating?: boolean,
+    }>,
+  }>,
+} = {};
```

Comment:
I feel like this should be a reducer of its own... it might feel a bit redundant because it's only used here, but I actually think it would help clarify the sequence of events.

In addition to some other thoughts I have, here's how I imagine the flow looking:

```
- install single dependency
- check to see if it's locked
- if not, dispatch an action, like `ADD_NEW_DEPENDENCY_START`
    - The notification reducer listens for that action and shows the "installing" notification
    - The queue reducer listens for that action and creates a new queue for this project
- If so, dispatch an action like `QUEUE_NEW_DEPENDENCY`
    - the notification reducer adds a new notification
    - The queue reducer does the compounding logic and such, and figures out the new queue

- The saga listens for finish events, and when a dependency update finishes, it selects the queue from the store to see what the next item is
```

I know this is a pretty big change I'm suggesting, but I think it would be a huge boon for being able to follow the sequence of events. Would be helpful for debugging / maintenance.

cc @mik639, I'm keen to get your opinion on this (as well as this entire file)